Add first implementation of the Emacs Lisp representer

fapdash opened the pull request:
Heavily inspired by the Clojure representer implementation by @bobbicodes.

## TODOs

- [ ] remove docstrings from [eieio](https://www.gnu.org/software/emacs/manual/html_node/eieio/) functions:
  - [ ]  `cl-defgeneric`
  - [ ] `cl-defmethod`
- [ ] download solutions for testing of representer efficiency
  - https://github.com/IsaacG/exercism-py/blob/612831462c2e1736fdc3e8ccb96958eac543b467/exercism.py#L65-L76
  - https://exercism.org/api/v2/tracks/emacs-lisp/exercises/two-fer/community_solutions

fapdash committed:
Add first implementation of the Emacs Lisp representer

Heavily inspired by the Clojure representer implementation by
@bobbicodes.

fapdash commented:
@BNAndras @kmarker1101 @keiravillekode Not finished yet, but working. Already tagging you so you can have a look and give feedback, if you'd like. :)

fapdash committed:
Extract method exercism//represent

fapdash committed:
Use threading macro, easier to read

fapdash committed:
Remove docstrings from function definitions

fapdash committed:
Remove docstrings from defvar and defconst

fapdash committed:
Add newline to end of file in .gitignore

fapdash committed:
Generate representation.json with newline at EOF

BNAndras commented:
I haven't built or used an Exercism representer before so I don't know how useful I'd be. Maybe @ErikSchierboom would be a good source of feedback

fapdash committed:
Remove line / statement not doing anything

fapdash committed:
Remove docstrings of function definitions after macroexpand

Could have done the docstring removal via another treepy-prewalk
instead of doing mutations. Not sure what I prefer.

fapdash committed:
Explain more in comment

ErikSchierboom commented:
@fapdash I've created the https://github.com/exercism/emacs-lisp-representer/ repo for you.

ErikSchierboom reviewed (approved):
Looks good!

fapdash commented:
I haven't had much time to work on this lately, but wanted to give a sign of life. Will hopefully pick this up again in a month or so.

fapdash committed:
Exclude compiled Elisp from Git

fapdash committed:
Start work on removal of docstrings
